src/Functions/stringToMap.cpp

```diff
@@ -0,0 +1,242 @@
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors args{
+            {"str_expr", &isStringOrFixedString<IDataType>, nullptr, "String or Fixed String"}
+        };
+        FunctionArgumentDescriptors optional_args{
+            {"pair_delim", &isString<IDataType>, isColumnConst, "const String"},
+            {"key_value_delim", &isString<IDataType>, isColumnConst, "const String"}
+        };
+
+        validateFunctionArgumentTypes(*this, arguments, args, optional_args);
+
+        DataTypePtr string_type = std::make_shared<DataTypeString>();
+        DataTypePtr string_type_nullable = std::make_shared<DataTypeNullable>(string_type);
+        DataTypes key_value_type{string_type, string_type_nullable};
+
+        return std::make_shared<DataTypeMap>(key_value_type);
+    }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    static void executeString(const ColumnString * col_str_in, ColumnPtr & col_res, size_t input_rows_count, char pair_delim, char key_value_delim)
+    {
+        String pair_delim_str(1, pair_delim);
+        
+        const ColumnString::Chars & in_vec = col_str_in->getChars();
+        const ColumnString::Offsets & in_offsets = col_str_in->getOffsets();
+        size_t size = in_offsets.size();
+
+        MutableColumnPtr keys_data = DataTypeString().createColumn();
+        MutableColumnPtr values_data = DataTypeString().createColumn();
+        MutableColumnPtr offsets = DataTypeNumber<IColumn::Offset>().createColumn();
+
+        keys_data->reserve(input_rows_count);
+        values_data->reserve(input_rows_count);
+        offsets->reserve(input_rows_count);
+
+        ColumnUInt8::MutablePtr values_null_map = ColumnUInt8::create();
+
+        ColumnString::Offset curr_res_offset = 0, curr_in_offset = 0;
+
+        std::vector<String> key_value_pairs;
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            key_value_pairs.clear();
+            const char * pos_in = reinterpret_cast<const char *>(&in_vec[curr_in_offset]);
+            size_t curr_size = in_offsets[i] - curr_in_offset - 1;
+            // splitting the string into key-value pairs
+            boost::split(key_value_pairs, std::string(pos_in, curr_size), boost::is_any_of(pair_delim_str));
+            // preparing the key, value columns for the map
+            for (size_t j = 0; j < key_value_pairs.size(); j++)
+            {
+                const std::string_view str_view = key_value_pairs[j];
+                size_t delim_pos = str_view.find(key_value_delim);
+                if (delim_pos != str_view.npos)
+                {
+                    keys_data->insert(str_view.substr(0, delim_pos));
+                    values_data->insert(str_view.substr(delim_pos + 1));
+                    values_null_map->insert(false);
```

Comment:
I wonder if we could do a `reserve()` call on `values_null_map` similar to l. 77ff based on a worst-case estimation? Even if the estimation is way off, it will (in average) be better repeated implicit resizes triggered by `insert()`.